src/service/parallax-impl.js

```diff
@@ -0,0 +1,215 @@
+   * in the document.
+   * @param {!Window} global
+   * @private
+   */
+  installParallaxHandlers_(global) {
+    const doc = global.document;
+    const viewport = viewportForDoc(doc);
+    const vsync = vsyncFor(global);
+
+    const elements = toArray(doc.querySelectorAll(`[${ATTR}]`));
+    const parallaxElements = elements.map(
+        e => new ParallaxElement(e, this.transform_));
+    const mutate =
+        this.parallaxMutate_.bind(this, parallaxElements, viewport);
+
+    viewport.onScroll(() => vsync.mutate(mutate));
+    mutate(); // initialize the elements with the current scroll position
+  }
+
+  /**
+   * Update each [amp-fx-parallax] element with the new scroll position.
+   * Notify any listeners.
+   * @param {!Array<!ParallaxElement>} elements
+   * @param {!./viewport-impl.Viewport} viewport
+   * @private
+   */
+  parallaxMutate_(elements, viewport) {
+    const newScrollTop = viewport.getScrollTop();
+    const previousScrollTop = this.getPreviousScroll_();
+    const delta = previousScrollTop - newScrollTop;
+
+    elements.forEach(element => {
+      if (!element.shouldUpdate(viewport)) {
+        return;
+      }
+      element.update(delta);
+      this.setPreviousScroll_(newScrollTop);
+    });
+
+    this.fire_(newScrollTop);
+  }
+
+  /**
+   * Create a value for the CSS transform property given a position.
+   * @param {number} position
+   * @return {string}
+   */
+  transform_(position) {
+    return `translate3d(0,${position.toFixed(2)}px,0)`;
+  }
+
+  /**
+   * Get the previous scroll value.
+   * @return {number}
+   * @private
+   */
+  getPreviousScroll_() {
+    return this.previousScroll_;
+  }
+
+  /**
+   * Set the previous scroll value.
+   * @param {number} scroll
+   * @private
+   */
+  setPreviousScroll_(scroll) {
+    this.previousScroll_ = scroll;
+  }
+
+  /**
+   * Add listeners to parallax scroll events.
+   * @param {!function()} cb
+   * @private
+   * @visibleForTesting
+   */
+  addScrollListener_(cb) {
+    this.parallaxObservable_.add(cb);
+  }
+
+  /**
+   * Remove listeners from parallax scroll events.
+   * @param {!function()} cb
+   * @private
+   * @visibleForTesting
+   */
+  removeScrollListener_(cb) {
+    this.parallaxObservable_.remove(cb);
+  }
+
+  /**
+   * Alert listeners that a scroll has occurred.
+   * @param {number} scrollTop
+   * @private
+   */
+  fire_(scrollTop) {
+    this.parallaxObservable_.fire(scrollTop);
+  }
+}
+
+/**
+ * Encapsulates and tracks an element's linear parallax effect.
+ */
+export class ParallaxElement {
+  /**
+   * @param {!Element} element The element to give a parallax effect.
+   * @param {!function(number):string} transform Computes the transform from the position.
+   */
+  constructor(element, transform) {
+    const factor = element.getAttribute(ATTR);
+
+    /** @private @const {!Element} */
+    this.element_ = element;
+
+    /** @private @const {!function(number):string} */
+    this.transform_ = transform;
```

Comment:
Why are we passing this in? Couldn't this just be on the `ParallaxElement` class?